File: server-mod/src/GiveStashItemSptCommand.ts
import {inject, injectable} from "tsyringe";
import {ISptCommand} from "@spt/helpers/Dialogue/Commando/SptCommands/ISptCommand";
import {ItemHelper} from "@spt/helpers/ItemHelper";
import {ISendMessageRequest} from "@spt/models/eft/dialog/ISendMessageRequest";
import {ISptProfile} from "@spt/models/eft/profile/ISptProfile";
import {IUserDialogInfo} from "@spt/models/eft/profile/IUserDialogInfo";
import {MailSendService} from "@spt/services/MailSendService";
import type {ICloner} from "@spt/utils/cloners/ICloner";
import {SaveServer} from '@spt/servers/SaveServer';
import {IItem} from "@spt/models/eft/common/tables/IItem";


interface InventoryItemHash {
    byItemId: Record<string, IItem>;
    byParentId: Record<string, IItem[]>;
}

@injectable()
export class GiveStashItemSptCommand implements ISptCommand {
    private static commandRegex = /^spt give-user-stash-item ((([a-z]{2,5}) )?"(.+)"|\w+)$/;

    public constructor(
        @inject("ItemHelper") protected itemHelper: ItemHelper,
        @inject("MailSendService") protected mailSendService: MailSendService,
        @inject("PrimaryCloner") protected cloner: ICloner,
        @inject("SaveServer") protected saveServer: SaveServer,
    ) {
    }

    public getCommand(): string {
        return "give-user-stash-item";
    }

    public getCommandHelp(): string {
        return 'spt give-user-stash-item\n========\nSends items to the player through the message system.\n\n\tspt give-user-stash-item [stashItem.Id]';
    }

    public performAction(commandHandler: IUserDialogInfo, sessionId: string, request: ISendMessageRequest): string {
        if (!GiveStashItemSptCommand.commandRegex.test(request.text)) {
            this.mailSendService.sendUserMessageToPlayer(
                sessionId,
                commandHandler,
                'Invalid use of give command. Use "help" for more information.',
            );
            return request.dialogId;
        }

        const result = GiveStashItemSptCommand.commandRegex.exec(request.text);

        if (!result[1]) {
            this.mailSendService.sendUserMessageToPlayer(
                sessionId,
                commandHandler,
                `Invalid use of give command. Use "help" for more information.`,
            );
            return request.dialogId;
        }

        const profile: ISptProfile = this.saveServer.getProfiles()[sessionId];

        const itemId = result[1];
        const inventoryItemHash = this.getInventoryItemHash(profile.characters.pmc.Inventory.items)

        const itemToAdd = inventoryItemHash.byItemId[itemId]
        if (!itemToAdd) {
            this.mailSendService.sendUserMessageToPlayer(
                sessionId,
                commandHandler,
                `Couldn't find item with id: ${itemId}`,
            );
            return request.dialogId;
        }


        const checkedItem = this.itemHelper.getItem(itemToAdd._tpl);
        if (!checkedItem[0]) {
            this.mailSendService.sendUserMessageToPlayer(
                sessionId,
                commandHandler,
                "That item could not be found. Please refine your request and try again.",
            );
            return request.dialogId;
        }
        // TODO
        // const quantity = checkedItem[1]._props.StackMaxSize

        const allChild = this.getAllDescendantsIncludingSelf(itemId, inventoryItemHash)
        let itemsToSend = this.cloner.clone(allChild);
        // console.log('original allChild', allChild);

        itemsToSend = this.itemHelper.replaceIDs(itemsToSend);
        this.itemHelper.setFoundInRaid(itemsToSend);
        this.mailSendService.sendSystemMessageToPlayer(sessionId, "SPT GIVE", itemsToSend);

        // console.log('itemsToSend final', itemsToSend);

        return request.dialogId;
    }

    private getInventoryItemHash(inventoryItem: IItem[]): InventoryItemHash {
        const inventoryItemHash: InventoryItemHash = { byItemId: {}, byParentId: {} };
        for (const item of inventoryItem) {
            inventoryItemHash.byItemId[item._id] = item;

            if (!("parentId" in item)) {
                continue;
            }

            if (!(item.parentId in inventoryItemHash.byParentId)) {
                inventoryItemHash.byParentId[item.parentId] = [];
            }
            inventoryItemHash.byParentId[item.parentId].push(item);
        }
        return inventoryItemHash;
    }

    private getAllDescendantsIncludingSelf(parentId: string, hash: InventoryItemHash): IItem[] {
        const result: IItem[] = [];

        const self = hash.byItemId[parentId];
        if (self) {
            result.push(self); // Add the original item first
        }

        const directChildren = hash.byParentId[parentId] || [];

        for (const child of directChildren) {
            result.push(...this.getAllDescendantsIncludingSelf(child._id, hash));
        }

        return result;
    }
}

File: server-mod/GiveUI/Command/GiveStashItemSptCommand.cs
using System.Text.RegularExpressions;
using SPTarkov.DI.Annotations;
using SPTarkov.Server.Core.Helpers;
using SPTarkov.Server.Core.Helpers.Dialog.Commando.SptCommands;
using SPTarkov.Server.Core.Models.Common;
using SPTarkov.Server.Core.Models.Eft.Common.Tables;
using SPTarkov.Server.Core.Models.Eft.Dialog;
using SPTarkov.Server.Core.Models.Eft.Profile;
using SPTarkov.Server.Core.Servers;
using SPTarkov.Server.Core.Services;
using SPTarkov.Server.Core.Utils.Cloners;

namespace GiveUI.Command;

[Injectable]
public class GiveStashItemSptCommand(
    MailSendService mailSendService,
    SaveServer saveServer,
    ICloner cloner,
    ItemHelper itemHelper) : ISptCommand
{
    private static readonly Regex _commandRegex = new(@"^spt give-user-stash-item ((([a-z]{2,5}) )?""(.+)""|\w+)$");

    public string GetCommand()
    {
        return "give-user-stash-item";
    }

    public string GetCommandHelp()
    {
        return
            "spt give-user-stash-item\n========\nSends items to the player through the message system.\n\n\tspt give-user-stash-item [stashItem.Id]";
    }

    public ValueTask<string> PerformAction(UserDialogInfo commandHandler, string sessionId, SendMessageRequest request)
    {
        if (!_commandRegex.IsMatch(request.Text))
        {
            mailSendService.SendUserMessageToPlayer(
                sessionId,
                commandHandler,
                "Invalid use of give command. Use 'help' for more information."
            );
            return new ValueTask<string>(request.DialogId);
        }

        var result = _commandRegex.Match(request.Text);
        var itemId = result.Groups[1].Value;
        if (string.IsNullOrEmpty(itemId))
        {
            mailSendService.SendUserMessageToPlayer(
                sessionId,
                commandHandler,
                "Invalid use of give command. Use 'help' for more information."
            );
            return new ValueTask<string>(request.DialogId);
        }

        var profile = saveServer.GetProfiles()[sessionId];

        var inventoryItemHash = GetInventoryItemHash(profile.CharacterData?.PmcData?.Inventory?.Items ?? []);
        var itemToAdd = inventoryItemHash.ByItemId[itemId];
        if (itemToAdd == null)
        {
            mailSendService.SendUserMessageToPlayer(
                sessionId,
                commandHandler,
                $"Couldn't find item with Id: {itemId}"
            );
            return new ValueTask<string>(request.DialogId);
        }

        var checkedItem = itemHelper.GetItem(itemToAdd.Template);
        if (!checkedItem.Key)
        {
            mailSendService.SendUserMessageToPlayer(
                sessionId,
                commandHandler,
                $"Couldn't find template with id: ${itemToAdd.Template}"
            );
            return new ValueTask<string>(request.DialogId);
        }

        var allChild = GetAllDescendantsIncludingSelf(itemId, inventoryItemHash);
        var itemsToSend = cloner.Clone(allChild) ?? [];

        itemsToSend = itemHelper.ReplaceIDs(itemsToSend, null);
        itemHelper.SetFoundInRaid(itemsToSend);
        mailSendService.SendSystemMessageToPlayer(sessionId, "SPT GIVE", itemsToSend);

        return new ValueTask<string>(request.DialogId);
    }

    private InventoryItemHash GetInventoryItemHash(List<Item> inventoryItems)
    {
        var inventoryItemHash = new InventoryItemHash {
            ByItemId = new Dictionary<MongoId, Item>(),
            ByParentId = new Dictionary<MongoId, HashSet<Item>>()
        };

        foreach (var item in inventoryItems)
        {
            inventoryItemHash.ByItemId[item.Id] = item;

            if (item.ParentId == null)
            {
                continue;
            }

            if (!inventoryItemHash.ByParentId.ContainsKey(item.ParentId))
            {
                inventoryItemHash.ByParentId[item.ParentId] = [];
            }

            inventoryItemHash.ByParentId[item.ParentId].Add(item);
        }

        return inventoryItemHash;
    }

    private List<Item> GetAllDescendantsIncludingSelf(string parentId, InventoryItemHash hash)
    {
        var result = new List<Item>();

        if (hash.ByItemId.TryGetValue(parentId, out var self))
        {
            result.Add(self);
        }

        if (hash.ByParentId.TryGetValue(parentId, out var directChildren))
        {
            foreach (var child in directChildren)
            {
                result.AddRange(GetAllDescendantsIncludingSelf(child.Id, hash));
            }
        }

        return result;
    }
}